ffi/src/table_changes.rs

```diff
@@ -0,0 +1,832 @@
+        commit_add_file(storage.as_ref(), 1, PARQUET_FILE2.to_string()).await?;
+
+        let batch = generate_batch_with_id(1)?;
+        put_file(storage.as_ref(), PARQUET_FILE1.to_string(), &batch).await?;
+        let batch = generate_batch_with_id(4)?;
+        put_file(storage.as_ref(), PARQUET_FILE2.to_string(), &batch).await?;
+
+        let path = "memory:///";
+        let engine = DefaultEngine::new(storage.clone(), Arc::new(TokioBackgroundExecutor::new()));
+        let engine = engine_to_handle(Arc::new(engine), allocate_err);
+
+        let table_changes = ok_or_panic(unsafe {
+            table_changes_from_version(kernel_string_slice!(path), engine.shallow_copy(), 0)
+        });
+
+        assert_eq!(
+            unsafe { table_changes_start_version(table_changes.shallow_copy()) },
+            0
+        );
+        assert_eq!(
+            unsafe { table_changes_end_version(table_changes.shallow_copy()) },
+            1
+        );
+
+        let table_root =
+            unsafe { table_changes_table_root(table_changes.shallow_copy(), allocate_str) };
+        assert_eq!(recover_string(table_root.unwrap()), path);
+
+        let schema = unsafe { table_changes_schema(table_changes.shallow_copy()).shallow_copy() };
+        let schema_ref = unsafe { schema.as_ref() };
+        assert_eq!(schema_ref.fields().len(), 5);
+        check_columns_in_schema(
+            &[
+                "id",
+                "val",
+                "_change_type",
+                "_commit_version",
+                "_commit_timestamp",
+            ],
+            schema_ref,
+        );
+
+        let table_changes_scan =
+            ok_or_panic(unsafe { table_changes_scan(table_changes, engine.shallow_copy(), None) });
+
+        let table_root = unsafe {
+            table_changes_scan_table_root(table_changes_scan.shallow_copy(), allocate_str)
+        };
+        assert_eq!(recover_string(table_root.unwrap()), path);
+
+        let logical_schema = unsafe {
+            table_changes_scan_logical_schema(table_changes_scan.shallow_copy()).shallow_copy()
+        };
+        let logical_schema_ref = unsafe { logical_schema.as_ref() };
+        assert_eq!(logical_schema_ref.fields().len(), 5);
+        check_columns_in_schema(
+            &[
+                "id",
+                "val",
+                "_change_type",
+                "_commit_version",
+                "_commit_timestamp",
+            ],
+            logical_schema_ref,
+        );
+
+        let physical_schema = unsafe {
+            table_changes_scan_physical_schema(table_changes_scan.shallow_copy()).shallow_copy()
+        };
+        let physical_schema_ref = unsafe { physical_schema.as_ref() };
+        assert_eq!(physical_schema_ref.fields().len(), 2);
+        check_columns_in_schema(&["id", "val"], physical_schema_ref);
+
+        unsafe {
+            free_table_changes_scan(table_changes_scan);
+            free_engine(engine);
+        }
```

Comment:
@kssenii  the latest build indicates that we're leaking memory for the schema:
```
507 | #[derive(Debug, PartialEq, Clone, Eq)]
    |                            ^^^^^
note: inside `table_changes::table_changes_schema`
   --> ffi/src/table_changes.rs:118:14
    |
118 |     Arc::new(table_changes.schema().clone()).into()
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
note: inside closure
   --> ffi/src/table_changes.rs:584:31
    |
584 |         let schema = unsafe { table_changes_schema(table_changes.shallow_copy()).shallow_copy() };

```

I think we may need to call `free_schema` in the test for the newly allocated copy of the schema.

```suggestion
        unsafe {
            free_table_changes_scan(table_changes_scan);
            free_engine(engine);
            free_schema(schema);
        }
```